qkit/measure/timedomain/pulse_sequence.py

```diff
@@ -46,73 +63,275 @@ def __call__(self, time_fractions):
+                readout_index = index
+            else:
+                if length > 0.5*timestep:
+                    wfm = pulse_dict["pulse"](np.arange(0, length, timestep) / length)
+                else:
+                    wfm = np.zeros(0)
+            
+            if (length < 0.5*timestep) and (length != 0):
+                logging.warning("{:}-pulse is shorter than {:.2f} nanoseconds and thus is omitted.".format(pulse_dict["name"], 0.5*timestep*1e9))
+
+            # Append wfm to waveform: Make sure waveform is long enough and account for skipping
+            wfm = np.atleast_1d(wfm)
+            waveform = np.append(waveform, np.zeros(max(0, len(wfm) - len(waveform) + index)))
+            waveform[index:index + len(wfm)] += wfm
+            if not pulse_dict["skip"]:
+                index = len(waveform)
+        
+        if readout_index == -1:
+            readout_index = len(waveform)
 
-        # Sorting the pulse start times makes generating the waveform less expensive
-        pulses = self.get_pulse_dict()
+        return waveform, readout_index
 
-        index = 0
-        for pulse_time, pulse in pulses.iteritems():
-            while times[index] < pulse_time:
-                index = index + 1
-            # assert: index points to first time where pulse starting at time t plays a role
+    def add(self, pulse, skip = False):
+        """
+        Add a pulse object to the sequence.
+        If skip is True the next pulse in the sequence will not wait until this pulse is finished (i.e. they happen at the same time).
+        """
+        pulse_dict = {}
+        if pulse.name in ["readout", "wait"]:
+            print(pulse.name + " is not permitted as pulse name.")
+            print("Pulse name set to None.")
+            pulse_dict["name"] = None
+        else:
+            pulse_dict["name"] = pulse.name
+        pulse_dict["shape"] = pulse.shape.name
+        if isinstance(pulse.length, float):
+            pulse_dict["length"] = pulse.length
+        elif callable(pulse.length):
+            pulse_dict["length"] = pulse.length
+            varnum = len(getargspec(pulse.length)[0])
+            if self._varnum is 0:
+                self._varnum = varnum
+            elif self._varnum is varnum:
+                pass
+            else:
+                print("Number of variable does not match previously added pulses/wait times!")
+                return self
+        else:
+            print("Pulse length not understood.")
+            return self
+        pulse_dict["pulse"] = pulse
+        pulse_dict["skip"] = skip
+        self._pulses.append(pulse_dict)
+        return self
 
-            if not isinstance(pulse, Pulse):
-                # Only a marker pulse -> not visible in waveform
-                continue
+    def add_wait(self, time):
+        """
+        Add a wait time to the sequence.
+        Using a (lambda) function makes the wait time variable.
+        """
+        pulse_dict = {}
```

Comment:
Code duplicity with add()...
Maybe move common code to extra private method
like  `_add_pulse(name, length, skip=False)`